### PR TITLE
Add support for NearMobileWallet 

### PR DESCRIPTION
**Pull request by GuillemGarciaDev**
# Description

Please include a summary of the wallet and how users can install it. Be sure to take a look at our guide on [Custom Wallets](https://github.com/near/wallet-selector/blob/main/packages/core/docs/guides/custom-wallets.md) to understand core concepts when integrating with Wallet Selector.

## Summary

In order to onboard customers with our wallet, this PR intends to connect NearMobileWallet with Wallet Selector. Users of the Near Mobile wallet may safely store, manage, and send NEAR and other cryptocurrencies while on the go. Non-custodial mobile wallet that is simple to use.

## PR summary
This PR contains the integration with Wallet selector. To achieve it, a package has been added to `package.json` file (`@peersyst/near-mobile-signer`) to handle the requests. 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by HarisSQA** [comment]
Hi @GuillemGarciaDev ,

I try to run this PR, and I can see wallet in list, but when I try to open it, I can't see QR code there.

**Comment by erditkurteshiSQA** [comment]
Hi @GuillemGarciaDev  ,
Thank you for raising this PR.

While trying to login with near mobile wallet we noticed a QR code shown in the modal.
Can you explain how to scan the QR code, on the mobile app there is no place to scan it.

image

**Comment by GuillemGarciaDev** [comment]
> Hi @GuillemGarciaDev , Thank you for raising this PR.
> 
> While trying to login with near mobile wallet we noticed a QR code shown in the modal. Can you explain how to scan the QR code, on the mobile app there is no place to scan it.
> 
> image

Hi @erditkurteshiSQA! 
We are preparing the version of the wallet with the new functionality. Once it is updated, you will be able to scan the QR using a scan button on the bottom bar.

When the version is uploaded I will notify you so you can test it.

Thanks for your quick reply!

**Comment by GuillemGarciaDev** [comment]
> Hi @GuillemGarciaDev ,
> 
> I try to run this PR, and I can see wallet in list, but when I try to open it, I can't see QR code there.

Hi @HarisSQA !

When you click on the wallet, does a modal appear or nothing at all?

Thanks for the feedback!

**Comment by HarisSQA** [comment]
> > Hi @GuillemGarciaDev ,
> > I try to run this PR, and I can see wallet in list, but when I try to open it, I can't see QR code there.
> 
> Hi @HarisSQA !
> 
> When you click on the wallet, does a modal appear or nothing at all?
> 
> Thanks for the feedback!

Hi @GuillemGarciaDev ,

I had an issue only with the QR code. Modal was working fine. 

Also, I have tested it now, and the QR code is shown in modal.

**Comment by GuillemGarciaDev** [comment]
> > > Hi @GuillemGarciaDev ,
> > > I try to run this PR, and I can see wallet in list, but when I try to open it, I can't see QR code there.
> > 
> > 
> > Hi @HarisSQA !
> > When you click on the wallet, does a modal appear or nothing at all?
> > Thanks for the feedback!
> 
> Hi @GuillemGarciaDev ,
> 
> I had an issue only with the QR code. Modal was working fine.
> 
> Also, I have tested it now, and the QR code is shown in modal.

Hello!

Sorry for taking so long to reply. @HarisSQA we are trying to replicate your case! Thanks for the feedback.

Possibly tomorrow or the day after you will have the new version of the app available to test the wallet connector with the wallet. @HarisSQA @erditkurteshiSQA 

When it's up I'll notify you!

Thanks again!

**Comment by GuillemGarciaDev** [comment]
Hi! @HarisSQA @erditkurteshiSQA 

The new version of NEAR Mobile wallet has been uploaded to the stores, so you can now test the PR properly. Once you import an account, you will find a scan button on the bottom bar of the app that will allow you to scan QRs.

I'll keep an eye out for any feedback.
Thanks again

**Comment by erditkurteshiSQA** [comment]
Hi @GuillemGarciaDev ,

We have tested the wallet with the latest release of the app.
Noticed a problem while scanning the qr code. 

Take a look at the picture below:

File: packages/near-mobile-wallet/assets/icon.png
Comment (erditkurteshiSQA) [inline packages/near-mobile-wallet/assets/icon.png:1]: Can you please make this image smaller ?
Also can you update the base64 too?

File: packages/near-mobile-wallet/jest.config.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/* eslint-disable */
+export default {
+  displayName: "ramper-wallet",
```
Comment (erditkurteshiSQA) [inline packages/near-mobile-wallet/jest.config.ts:3]: Can you change the names to near mobile wallet in this file ?

File: packages/near-mobile-wallet/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/near/wallet-selector.git"
+  },
+  "bugs": {
+    "url": "https://github.com/near/wallet-selector/issues"
+  },
+  "homepage": "https://github.com/near/wallet-selector/tree/main/packages/near-mobile-wallet",
+  "peerDependencies": {
+    "near-api-js": "^1.0.0"
```
Comment (erditkurteshiSQA) [inline packages/near-mobile-wallet/package.json:25]: Can you set the range version to the same as other packages of near-api-js?

File: packages/near-mobile-wallet/project.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      "options": {
+        "jestConfig": "packages/near-mobile-wallet/jest.config.ts",
+        "passWithNoTests": true
+      }
+    },
+    "deploy": {
+      "executor": "ngx-deploy-npm:deploy",
+      "options": {
+        "access": "public"
+      }
```
Comment (erditkurteshiSQA) [inline packages/near-mobile-wallet/project.json:51]: Can you add here `"dependsOn": ["^deploy"]`?
See other packages how they implemented.

File: packages/near-mobile-wallet/assets/icon.png
Comment (GuillemGarciaDev) [reply packages/near-mobile-wallet/assets/icon.png:1]: Base64 version is already in the project  at `near-mobile-wallet/src/lib/icon.ts`

**Comment by GuillemGarciaDev** [comment]
> Hi @GuillemGarciaDev ,
> 
> We have tested the wallet with the latest release of the app. I tested it with importing two different accounts and also creating one. Noticed a problem while scanning the QR code.
> 
> Take a look at the picture below:
> 
> 

Hi @erditkurteshiSQA!

The issue is now fixed! It was a missing env variable. I tested the guest-book locally with the app and it seems to be working fine. 

All requested PR changes have been fixed!

Regards, Guillem

**Comment by erditkurteshiSQA** [comment]
Hi @GuillemGarciaDev ,
Thank you for the quick fix.

We were able to log in with iOS and tested it. However, the `signMessage` feature is not working, and I will provide a picture below of the error:

image

Also, we noticed that if we close the QR modal while trying to connect with the wallet, we still see the message `Connecting to Near Mobile Wallet` instead of a different message or an error message indicating that the modal is closed.

image

On Android, however, on the Play Store, it's still the older version of the wallet, and the login problem still persists.

Play Store app version: 1.3.0
App Store app version: 2.3.0

File: packages/near-mobile-wallet/jest.config.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  transform: {
+    "^.+\\.[tj]s$": [
+      "ts-jest",
+      {
+        tsconfig: "<rootDir>/tsconfig.spec.json",
+      },
+    ],
+  },
+  moduleFileExtensions: ["ts", "tsx", "js", "jsx"],
+  coverageDirectory: "../../coverage/packages/ramper-wallet",
```
Comment (erditkurteshiSQA) [inline packages/near-mobile-wallet/jest.config.ts:15]: Can you please update the directory to near-mobile-wallet?

**Comment by GuillemGarciaDev** [comment]
Hi @HarisSQA !

We have been working on the Android issue and it is already solved. You will be able to test it when it is in Play store. 👍🏻 

On the other hand, could you provide me some info about how the signed message is verified? What we do is to verify the signature using the public key of the account.

```typescript
const { publicKey, signature, accountId } = response;
const sign = new Uint8Array(Buffer.from(signature, "base64"));
const messageHash = new Uint8Array(sha256.digest(`NEP0413:${JSON.stringify({ receiver, message })}`));
const isValidSignature = PublicKey.fromString(publicKey).verify(messageHash, sign);
```

Thank you for your feedback!

**Comment by erditkurteshiSQA** [comment]
Hi @GuillemGarciaDev ,

Please refer to the example referenced here in the NEP:
https://github.com/near/NEPs/blob/master/neps/nep-0413.md#references

This is how the message needs to be serialized and signed in the wallet:
https://github.com/gagdiez/near-login/blob/main/tests/authentication/wallet.ts#L60

This is how we verify on our examples:
https://github.com/near/wallet-selector/blob/main/examples/react/components/Content.tsx#L289

Looks like you have implemented some kind of an old version of the sign message looking at this line of code:
const messageHash = new Uint8Array(sha256.digest(`NEP0413:${JSON.stringify({ receiver, message })}`));

The NEP413 does not add a prefix string NEP0413 to the message now it uses a tag (number) there's no receiver either please see the interface:
https://github.com/near/NEPs/blob/master/neps/nep-0413.md#input-interface

Message must be signed with the above structure with params ordered alphabetically for the required params.

**Comment by GuillemGarciaDev** [comment]
> Hi @GuillemGarciaDev ,
> 
> Please refer to the example referenced here in the NEP: https://github.com/near/NEPs/blob/master/neps/nep-0413.md#references
> 
> This is how the message needs to be serialized and signed in the wallet: https://github.com/gagdiez/near-login/blob/main/tests/authentication/wallet.ts#L60
> 
> This is how we verify on our examples: https://github.com/near/wallet-selector/blob/main/examples/react/components/Content.tsx#L289
> 
> Looks like you have implemented some kind of an old version of the sign message looking at this line of code: const messageHash = new Uint8Array(sha256.digest(`NEP0413:${JSON.stringify({ receiver, message })}`));
> 
> The NEP413 does not add a prefix string NEP0413 to the message now it uses a tag (number) there's no receiver either please see the interface: https://github.com/near/NEPs/blob/master/neps/nep-0413.md#input-interface
> 
> Message must be signed with the above structure with params ordered alphabetically for the required params.

There's a new version in the Play Store to try out wallet selector in Android! There's a new update on the `signMessage` feature and it is now working!

**Comment by erditkurteshiSQA** [comment]
Hi @GuillemGarciaDev ,

We checked the latest iOS and android version and the functionalities are working including `signMessage` but
there are some issues/improvement that require attention:

1) All Reject(s) are being ignored, and the promise never resolves.
2) Open the wallet-selector modal and click on NearMobileWallet to open the QR code modal. However, closing the QR code modal does not trigger the expected rejection of the signIn request.
3) When signing in, trying `signAndSendTransaction(s)` and scanning the QR code, the rejection on the wallet does not trigger any action. It should reject the `signAndSendTransaction` by throwing an error.
4) Add the ability for local signing for FunctionCall transactions without a deposit. Currently, the wallet creates a Limited Access Key (LAK) but does not utilize it.
5) Multiple-transactions should be able to be signed just by signing the QR code once.

As a suggestion consider comparing the UX of wallet-selector with `Here wallet`, since both wallets are mobile wallets.

**Comment by AdriaCarrera** [comment]
Hi @erditkurteshiSQA and @HarisSQA it's Adrià from Peersyst team, thanks for all this feedback.

We have almost all the points that you mentioned done, except the 4th one. Is it something that you require to be done before getting approved or we can merge without this point and make an improvement later on? Also could you point us to some links or docs that better explain the local signing?

Thank you!

**Comment by erditkurteshiSQA** [comment]
Hey @AdriaCarrera 

The 4th point is being mentioned as an improvement and can be done later on once the other points are working fine.

The idea is to be able to sign FunctionCall transactions that don't have any deposit without confirming it on the Wallet since the FunctionCall Access-key (limited access key LAK) is on the localStorage of when we sign in with NearMobileWallet, this improves the UX.

Looks like you save the LAK in `near-mobile-signer:session`

For instance, one use case of such a transaction is the ability to send messages in our guest book example without needing confirmation in the wallet when no deposit is attached.

This approach is also implemented in `near-api-js` as demonstrated here:
https://github.com/near/near-api-js/blob/master/packages/wallet-account/src/wallet_account.ts#L306

In our wallet-connect package, we have a simpler version, as shown in here:
https://github.com/near/wallet-selector/blob/main/packages/wallet-connect/src/lib/wallet-connect.ts#L558-L566

The exact implementation might differ but the core idea is the same, I believe a similar approach can be implemented in the @peersyst/near-mobile-signer package.

If you need any more information or have more questions, feel free to reach out.

**Comment by erditkurteshiSQA** [comment]
Hi @GuillemGarciaDev  @AdriaCarrera ,

Following the recent update on the Near mobile wallet and this PR, we tested it on both iOS and Android. 
We are unable to login in both platforms. Please refer to the attached image for the error:
image


Additionally, kindly remember that the 4th point, mentioned earlier, is considered an improvement and can be addressed once the other points are resolved.

Thank you!

**Comment by GuillemGarciaDev** [comment]
> Hi @GuillemGarciaDev @AdriaCarrera ,
> 
> Following the recent update on the Near mobile wallet and this PR, we tested it on both iOS and Android. We are unable to login in both platforms. Please refer to the attached image for the error: image
> 
> Additionally, kindly remember that the 4th point, mentioned earlier, is considered an improvement and can be addressed once the other points are resolved.
> 
> Thank you!

Hi @erditkurteshiSQA !!

This issue seems already fixed and you have available a new version with all the improvements (except the 4th one). I hope you find this fixes working as expected!

Thank you again!!

**Review by erditkurteshiSQA — APPROVED** [review]

**Comment by erditkurteshiSQA** [comment]
@GuillemGarciaDev 

After testing the wallet with the latest version in both platform, all issues have been successfully resolved. We are now ready to merge this PR.
**A release is planned for tomorrow, which will include the changes from this PR.**

Thank you!

**Comment by kujtimprenkuSQA** [comment]
Hi, @GuillemGarciaDev & @AdriaCarrera we have published  a new version of the wallet-selector:

Release notes: https://github.com/near/wallet-selector/releases/tag/v8.7.0

Near Mobile Wallet: https://www.npmjs.com/package/@near-wallet-selector/near-mobile-wallet

**Comment by kujtimprenkuSQA** [comment]
Hi, @GuillemGarciaDev & @AdriaCarrera could you please take a look at the issues that need improvement:

- The `@peersyst/near-mobile-signer` has too many unnecessary dependencies: 
https://www.npmjs.com/package/@peersyst/near-mobile-signer?activeTab=dependencies
Can you please improve the bundler to include only the necessary dependencies and remove all `devDependencies` from NPM? - they get added to dApps which is not ideal.

- The QR Code of NearMobileWallet does not open correctly when other extensions are installed such as MetaMask, Nightly, Sender etc see below video:

[qr-code-issue-near-mobile-wallet.webm](https://github.com/near/wallet-selector/assets/95851345/56849111-c1c9-4958-9dde-b5ad10db5334)

**Comment by GuillemGarciaDev** [comment]
> Hi, @GuillemGarciaDev & @AdriaCarrera could you please take a look at the issues that need improvement:
> 
> * The `@peersyst/near-mobile-signer` has too many unnecessary dependencies:
>   https://www.npmjs.com/package/@peersyst/near-mobile-signer?activeTab=dependencies
>   Can you please improve the bundler to include only the necessary dependencies and remove all `devDependencies` from NPM? - they get added to dApps which is not ideal.
> * Passing a `callbackUrl` for `signMessage`params is failing can you take a look at it too?
> * The QR Code of NearMobileWallet does not open correctly when other extensions are installed such as MetaMask, Nightly, Sender etc see below video:
> 
>  qr-code-issue-near-mobile-wallet.webm

Hi! @kujtimprenkuSQA we are working to fix these issues.

About `callbackUrl` issue, since `NearMobileWallet` is an `InjectedWallet` there's no reason to use a callbackUrl because the signature action is done in a mobile app. 

Should we create a new PR with the new fixes?

Thank you for your feedback!

**Comment by kujtimprenkuSQA** [comment]
> > Hi, @GuillemGarciaDev & @AdriaCarrera could you please take a look at the issues that need improvement:
> > 
> > * The `@peersyst/near-mobile-signer` has too many unnecessary dependencies:
> >   https://www.npmjs.com/package/@peersyst/near-mobile-signer?activeTab=dependencies
> >   Can you please improve the bundler to include only the necessary dependencies and remove all `devDependencies` from NPM? - they get added to dApps which is not ideal.
> > * Passing a `callbackUrl` for `signMessage`params is failing can you take a look at it too?
> > * The QR Code of NearMobileWallet does not open correctly when other extensions are installed such as MetaMask, Nightly, Sender etc see below video:
> > 
> > qr-code-issue-near-mobile-wallet.webm
> 
> Hi! @kujtimprenkuSQA we are working to fix these issues.
> 
> About `callbackUrl` issue, since `NearMobileWallet` is an `InjectedWallet` there's no reason to use a callbackUrl because the signature action is done in a mobile app.
> 
> Should we create a new PR with the new fixes?
> 
> Thank you for your feedback!

Yes, a new PR should be made once the QR Code issue is fixed and also the dependencies for `@peersyst/near-mobile-signer`

**Pull request merged**